build/commands/scripts/sync.js

```diff
@@ -41,40 +49,189 @@ const maybeInstallDepotTools = (options = config.defaultOptions) => {
+  let args = [
+    'sync', '--nohooks', '--reset', '--revision',
+    'src@' + requiredChromiumRef, '--with_tags',
+    '--with_branch_heads', '--upstream'
+  ];
+
+  const syncWithForce = program.init || program.force
+  if (syncWithForce) {
+    args.push('--force')
+  }
+  if (program.cleanup) {
+    if (util.isGitExclusionExists(config.srcDir, 'brave/')) {
+      args.push('-D')
+    } else {
+      Log.warn(
+          '--cleanup is ignored, because brave dir is not excluded from src')
+    }
+  }
+
+  const latestSuccessfulSyncFile =
+      path.join(config.rootDir, '.brave_latest_successful_sync.json')
+  const latestSuccessfulSync = util.readJSON(latestSuccessfulSyncFile)
+  const expectedSuccessfulSync = {
+    chromiumRef: requiredChromiumRef,
+    gClientTimestamp: fs.statSync(config.gClientFile).mtimeMs.toString()
+  }
+
+  const chromiumNeedsUpdate =
+      shouldUpdateChromium(latestSuccessfulSync, expectedSuccessfulSync)
+  const shouldSyncSrc = chromiumNeedsUpdate || syncWithForce ||
+      program.sync_src || program.cleanup
+  if (!shouldSyncSrc) {
+    return false
+  }
+
+  if (program.sync_src !== undefined && !program.sync_src) {
+    Log.warn(chalk.yellow.bold(
+        'Chromium needed sync but received the flag to skip performing the ' +
+        'update. Working directory may not compile correctly.'))
+    return false
+  }
+
+  util.runGClient(args, {cwd: config.rootDir}, config.gClientFile)
+  util.addGitExclusion(config.srcDir, 'brave/')
+  util.writeJSON(latestSuccessfulSyncFile, expectedSuccessfulSync)
+
+  const postSyncChromiumRef = util.getGitReadableLocalRef(config.srcDir)
+  Log.status(`Chromium is now at ${postSyncChromiumRef || '[unknown]'}`)
+  return true
+}
+
+function syncBrave(program) {
+  let args = ['sync', '--nohooks']
+  const syncWithForce = program.init || program.force
+  if (syncWithForce) {
+    args.push('--force')
+  }
+  if (program.cleanup) {
+    args.push('-D')
+  }
+
+  // Don't pass gClientFile here, let gclient find it automatically, which
+  // should be brave/.gclient.
+  util.runGClient(args, {cwd: config.braveCoreDir}, null)
+}
+
+function gclientRunhooks() {
+  util.runGClient(['runhooks'], {cwd: config.rootDir}, config.gClientFile)
+}
+
 async function RunCommand () {
   program.parse(process.argv)
   config.update(program)
 
-  if (program.all || program.run_hooks || program.run_sync) {
-    Log.warn('--all, --run_hooks and --run_sync are deprecated. Will behave as if flag was not passed. Please update your command to `npm run sync` in the future.')
-  }
-
   if (program.init || !fs.existsSync(config.depotToolsDir)) {
     maybeInstallDepotTools()
   }
 
-  if (program.init) {
-    util.buildGClientConfig()
+  if (program.init || !fs.existsSync(config.defaultGClientFile)) {
+    buildDefaultGClientConfig()
+  } else if (program.target_os) {
+    Log.warn(
+        '--target_os is ignored. Use with --init or edit .gclient manually')
```

Comment:
I didn't make this as an error, because android devs may always call sync with `--target_os android` which seems like an okay thing, no need to stop the process. The idea for the warning is to show that it should be used with `--init`, otherwise it doesn't do anything.